File: solutions/cpp/protein-translation/1/protein_translation.cpp
// Code dedicated to Shree DR.MDD
#include <unordered_map>
#include "protein_translation.h"
namespace protein_translation {
  static const auto codon_table = unordered_map<string, string>{
    { "AUG", "Methionine" },
    { "UUU", "Phenylalanine" },
    { "UUC", "Phenylalanine" },
    { "UUA", "Leucine" },
    { "UUG", "Leucine" },
    { "UCU", "Serine" },
    { "UCC", "Serine" },
    { "UCA", "Serine" },
    { "UCG", "Serine" },
    { "UAU", "Tyrosine" },
    { "UAC", "Tyrosine" },
    { "UGU", "Cysteine" },
    { "UGC", "Cysteine" },
    { "UGG", "Tryptophan" },
    { "UAA", "STOP" },
    { "UAG", "STOP" },
    { "UGA", "STOP" },
  };

  auto split_codons(const string& sequence) -> vector<string> {
    vector<string> codons;
    for (size_t idx = 0; idx + 2 < sequence.length(); idx += 3) {
      codons.emplace_back(sequence.substr(idx, 3));
    }
    return codons;
  }

  auto proteins(const string& rna_sequence) -> vector<string> {
    vector<string> chain;
    for (const auto& codon : split_codons(rna_sequence)) {
      const auto& product = codon_table.at(codon);
      if (product == "STOP") break;
      chain.emplace_back(product);
    }
    return chain;
  }
}

File: solutions/euphoria/protein-translation/1/protein-translation.ex
include std/sequence.e

public function proteins(sequence strand)
  sequence codons = breakup(strand, 3)
  sequence res = {}
  for i = 1 to length(codons) do
    if equal(codons[i], "AUG") then
      res &= {"Methionine"}
    elsif equal(codons[i], "UUU") or equal(codons[i], "UUC") then
      res &= {"Phenylalanine"}
    elsif equal(codons[i], "UUA") or equal(codons[i], "UUG") then
      res &= {"Leucine"}
    elsif equal(codons[i], "UCU") or equal(codons[i], "UCC") or equal(codons[i], "UCA") or equal(codons[i], "UCG") then
      res &= {"Serine"}
    elsif equal(codons[i], "UAU") or equal(codons[i], "UAC") then
      res &= {"Tyrosine"}
    elsif equal(codons[i], "UGU") or equal(codons[i], "UGC") then
      res &= {"Cysteine"}
    elsif equal(codons[i], "UGG") then
      res &= {"Tryptophan"}
    elsif equal(codons[i], "UAA") or equal(codons[i], "UAG") or equal(codons[i], "UGA") then
      return res
    else
      return {}
    end if
  end for
  return res
end function